tez-plugins/tez-aux-services/src/main/java/org/apache/tez/auxservices/ShuffleHandler.java

```diff
@@ -306,21 +305,24 @@ public ReduceMapFileCount(ReduceContext rc) {
 
     @Override
     public void operationComplete(ChannelFuture future) throws Exception {
+      Channel ch = future.channel();
       if (!future.isSuccess()) {
-        future.channel().close();
+        ch.close();
         return;
       }
       int waitCount = this.reduceContext.getMapsToWait().decrementAndGet();
       if (waitCount == 0) {
+        LOG.debug("Finished with all map outputs");
+        ch.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
```

Comment:
can we add comment why this is required here.

Reply:
ack, this is the most important part of this patch, added a commit with a code comment

Reply:
Is this issue mainly due to nettty upgrade (4.x?)

Reply:
yes, absolutely, this issue is because of the incorrect usage of netty4 APIs
most interestingly, there were no unit tests that showed this issue so far (added one now), which reproduces when we fetch more inputs in the same request: due to this issue, the new UT completely hung, and a real TPCDS query on the cluster became very slow, as composite fetch requests hung and timed out eventually (didn't cause a query failure, just an extremely slow query)